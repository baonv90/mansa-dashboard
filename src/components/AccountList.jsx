import React, { useCallback } from 'react'
import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  border-radius: .5rem;
  padding: .5rem;
  margin: 0 .5rem;
  background: ${({ isSelected }) => isSelected ?  'var(--primary-color)' : 'var(--grey-color)'};
  color: ${({ isSelected }) => isSelected ? 'white' : 'var(--text-color)'};
  min-width: 120px;
  min-height: 60px;
  cursor: pointer;
  align-items: center;
  justify-content: center;
`;

const AccountType = styled.span`
  font-size: .6rem;
  line-height: .6rem;
  font-weight: 400;
`;

const AccountNumber = styled.span`
  font-weight: 600;
`;


const AccountList = ({ isSelected, setSelectedTab, index, ...data }) => {
  const { account_number, account_type } = data;

  const onClick = useCallback(() => {
    setSelectedTab(index)
  }, [setSelectedTab, index]);

  return (
    <Container
      onClick={onClick}
      isSelected={isSelected}
    >
      <AccountType>{account_type}</AccountType>
      <AccountNumber>{account_number}</AccountNumber>
    </Container>
  )
}

export default AccountList