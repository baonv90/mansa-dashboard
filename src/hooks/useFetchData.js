import { useState, useEffect } from 'react';

// A hook to fetch data from given url
// it returns the fetching status loading and data
export const useFetchData = (url) => {
  const [data, setData] = useState();
  const [loading, setLoading] = useState(false);
  const [hasError, setHasError] = useState(false);

  useEffect(() => {
    setLoading(true);
    fetch(url).then((res) => {
      res.json().then((data) => {
        setData(data);
        setLoading(false);
      })
      .catch(() => {
        setHasError(true);
        setLoading(false);
      });
    });
  }, [url]);

  return { data, loading, hasError };
};