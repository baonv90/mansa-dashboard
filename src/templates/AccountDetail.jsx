import React from 'react'
import styled from 'styled-components';

import Block from '../layout/Block';
import BalanceChart from '../components/BalanceChart';
import Transactions from '../components/Transactions';
import { useFetchData } from '../hooks/useFetchData';
import { data_source } from './AccountInfo';

// For testing purpose, the start and end dates are fixed
const from = new Date('01 January 2019 00:00 UTC').toISOString();
const to = new Date('01 June 2019 00:00 UTC').toISOString();

const Container = styled.div`
  display: flex;
  margin-top: 1.5rem;
  max-height: 60vh;
`;

const BalanceWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 35%;
`;

const TransactionsWrapper = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  margin-left: 1rem;
`;

const Balance = styled.div`
  display: flex;
  align-items: baseline;
  margin-top: 1rem;
`;

const AccountBalance = styled.span`
  font-size: 2rem;
  line-height: 1.5rem;
  font-weight: 800;
`;

const Currency = styled.span`
  font-size: .8rem;
  line-height: .8rem;
  font-weight: 400;
  margin-left: .25rem;
`;

// Component to display current balance, balance chart, transaction history of one account
const AccountDetail = (props) => {
  const { available, currency, account_id } = props;
  const { data } = useFetchData(data_source + '/' + account_id + '/transactions?from=' + from + '&to=' + to);

  if (!data) {
    return null;
  }

  const chartData = (data || []).map(dat => dat.amount);

  return (
    <Container>
      <BalanceWrapper>
        <Block title='Current Balance'>
          <Balance>
            <AccountBalance>{available}</AccountBalance>
            <Currency>{currency}</Currency>
          </Balance>
        </Block>
        {chartData && chartData.length && <BalanceChart data={chartData} />}
      </BalanceWrapper>
      <TransactionsWrapper>
        <Block title='Transactions'>
          <Transactions data={data || []}/>
        </Block>
      </TransactionsWrapper>
    </Container>
  );
};

export default AccountDetail;
