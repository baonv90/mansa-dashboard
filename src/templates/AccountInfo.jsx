import React from 'react';
import styled from 'styled-components';

import Tabs from '../layout/Tabs';
import AccountDetail from './AccountDetail';
import AccountList from '../components/AccountList';
import { useFetchData } from '../hooks/useFetchData';

export const data_source = 'https://kata.getmansa.com/accounts';

const Container = styled.div`
  display: flex;
  flex-direction: column;
`;

// Component to display all accounts and its detail
const AccountInfo = () => {
  const { data } = useFetchData(data_source);

  if (!data) {
    return null;
  }

  return (
    <Container>
      <Tabs TabNav={AccountList}>
        {
          (data || []).map((accountData, index) => 
            <AccountDetail key={index} {...accountData} />
          )
        }
      </Tabs>
    </Container>
  )
};

export default AccountInfo;
