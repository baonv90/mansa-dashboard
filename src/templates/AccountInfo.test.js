import renderer from 'react-test-renderer';
import AccountInfo from './AccountInfo';
import AccountList from '../components/AccountList';
import { useFetchData } from '../hooks/useFetchData';
import AccountDetail from './AccountDetail';

jest.mock('../hooks/useFetchData');
jest.mock('./AccountDetail');


describe('templates > AccountInfo', () => {
  const fakeData = [{
      account_id: '1'
    }, {
      account_id: '2'
    }, {
      account_id: '3'
    }
  ];

  beforeEach(() => {
    useFetchData.mockReturnValue({
      loading: false,
      data: fakeData
    });

    AccountDetail.mockReturnValue(() => <div>test</div>);
  });

  it('should match snapshot', async () =>  {
    const component = await renderer.create(
      <AccountInfo />,
    );
    
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should pass data correctly to Transactions', async () =>  {
    const tree = await renderer.create(
      <AccountInfo />,
    );
    
    const tabs = tree.root.findAllByType(AccountList);
    expect(tabs).toHaveLength(fakeData.length);
  });
});
