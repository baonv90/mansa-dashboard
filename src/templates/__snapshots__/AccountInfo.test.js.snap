// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templates > AccountInfo should match snapshot 1`] = `
<div
  className="sc-jSFjdj kRAgvN"
>
  <div
    className="sc-bdnxRM fvtCJR"
  >
    <div
      className="sc-gtsrHT fSDNVI"
    >
      <div
        className="sc-dlnjwi hsVvGm"
        onClick={[Function]}
      >
        <span
          className="sc-hKFxyN fXoShm"
        />
        <span
          className="sc-eCApnc fxdTbP"
        />
      </div>
      <div
        className="sc-dlnjwi bGMjrZ"
        onClick={[Function]}
      >
        <span
          className="sc-hKFxyN fXoShm"
        />
        <span
          className="sc-eCApnc fxdTbP"
        />
      </div>
      <div
        className="sc-dlnjwi bGMjrZ"
        onClick={[Function]}
      >
        <span
          className="sc-hKFxyN fXoShm"
        />
        <span
          className="sc-eCApnc fxdTbP"
        />
      </div>
    </div>
  </div>
</div>
`;
